many: return kr/pretty:Diff when Equals/DeepEquals check fails

mvo5 opened the pull request:
This is a version of https://github.com/go-check/check/pull/99 that uses github.com/kr/pretty to do the diff formating. I don't mind if we use kr/pretty or  github.com/davecgh/go-spew/spew and github.com/pmezard/go-difflib/difflib but I would really like to see diff functionality in check.v1.

Here is an example of a non-trivial output:
```
FAIL: snapstate_test.go:1670: snapmgrTestSuite.TestInstallRunThrough

snapstate_test.go:1774:
    c.Assert(s.fakeBackend.ops, DeepEquals, expected)
... obtained snapstate_test.fakeOps = snapstate_test.fakeOps{snapstate_test.fakeOp{op:"storesvc-snap-action", name:"", channel:"", revno:snap.Revision{N:0}, sinfo:snap.SideInfo{RealName:"", SnapID:"", Revision:snap.Revision{N:0}, Channel:"", 
[16K output removed]

... Values are different, diff:
[1].userID: 1 != 2
[3].name: "some-snap" != "some-snapx"
[9].sinfo.Revision.N: 11 != 12
```

Thanks for your consideration and kudos for bboozzoo for the initial PR.

bboozzoo committed:
utils: add a helper for generating a diff between unequal values

Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>

bboozzoo committed:
checkers: return a pretty format error message in Equals and DeepEquals checkers

Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>

bboozzoo committed:
checkers: clear error message if Not checker sub test failed

Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>

bboozzoo committed:
utils: keep diff output under control

Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>

bboozzoo committed:
check_test: report regex parsing errors

Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>

bboozzoo committed:
checkers, fixture, foundation: update tests to handle diffs

Update tests to handle diff output in error message from Equals and DeepEquals.

Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>

mvo5 committed:
many: use github.com/kr/pretty for the diff generation

mvo5 commented:
Hey, sorry for naging. Is there anything I can do to help moving this (or #99) forward? It would be really nice to have a diff output when complex "check.Equal" (or DeepEqual) compares fail.

niemeyer commented on checkers.go:
```diff
@@ -90,6 +90,10 @@ func (checker *notChecker) Info() *CheckerInfo {
 	result = !result
+	if result {
+		// clear error message if the new result is true
+		error = ""
```
Is there a test for this?

mvo5 replied:
There was only an implicit test for this, I added an explicit one now.

niemeyer commented on fixture_test.go:
```diff
@@ -402,7 +402,9 @@ func (s *FixtureS) TestSetUpSuiteCheck(c *C) {
-			"\\.+ expected bool = true\n\n")
+			"\\.+ expected bool = true\n"+
+			"\\.+ Values are different, diff:\n"+
+			"false \\!= true\n\n")
```
This feels somewhat redundant (we have four lines saying true != false). We need to be a bit more strict about when the logic kicks in. If the goal is having typing, we might add additional logic inside the existing logValue function.

mvo5 replied:
Thats a very good point. I wonder what we should do here. We could simply omit this for all simple types. I.e. we do not show the difference for everything in `type.Basic. WDYT?

mvo5 replied:
As pointed out in a lot of cases this difference just adds noise. I updated the code to only do a diff if the type is complex and it wil also do very basic string diffs based on your suggestion to convert those into []string slices.

niemeyer commented on utils.go:
```diff
@@ -0,0 +1,20 @@
+func formatUnequal(obtained interface{}, expected interface{}) string {
+	diff := pretty.Diff(obtained, expected)
+	if len(diff) == 0 {
+		diff = []string{fmt.Sprintf("%p != %p", expected, obtained)}
```
This looks suspect. It doesn't seem better than what we already do for that case.

What's the background?

mvo5 replied:
I removed this again, this was to make it more inline with what #99 was doing (which uses a different diff library). But I agree that this adds no information.

niemeyer commented on utils.go:
```diff
@@ -0,0 +1,20 @@
+	}
+
+	return fmt.Sprintf(`Values are different, diff:
+%s`, strings.Join(diff, "\n"))
```
The prefix here might be simply "Difference:", and the printing should use the same logic we use in C.logMultiLine so content is properly indented and similar to the rest (see both that and logValue, inside check.go).

Also, this might be inside checkers.go next to the use cases. It's too simple to be on its own file.

mvo5 replied:
Thanks, the code uses the logMultiLine logic now and is moved into checkers.go next to where it is used.

niemeyer reviewed (changes requested):
Thanks for the changes. A few questions:

mvo5 committed:
Use logMultiLine logic to indent equal differences

This also moves the formatUnequal helper from utils.go into
checkers.go (as suggested during the code review).

mvo5 committed:
checkers_test.go: add explicit test for notChecker clearning the error string

mvo5 committed:
rework output of formatUnequal a bit, add integration tests

mvo5 committed:
show difference only for non-basic types, show multi-line string "diff"

bboozzoo reviewed (approved):
LGTM

mvo5 commented:
@bboozzoo Thanks for your review! Happy to tweak further if @niemeyer preferes the output to be different and/or more types should be included etc.

mvo5 commented on integration_test.go:
```diff
@@ -0,0 +1,95 @@
+... expected check_test.complexStruct = check_test.complexStruct{r:3, i:4}
+... Difference:
+...     r: 1 != 3
+ +
```
Gustavo mentioned on irc that this `+` looks out of place and I agree - I will look what I can do about it.

mvo5 committed:
check: fix formatMultiLine() to only add "+" when using quoted mode
